File: netlify/functions/webhook.js
const TELEGRAM_BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const VELLUM_API_KEY = process.env.VELLUM_API_KEY;
const VELLUM_WORKFLOW_NAME = process.env.VELLUM_WORKFLOW_NAME || "telegram-split-bill-bot";

const TELEGRAM_API = `https://api.telegram.org/bot${TELEGRAM_BOT_TOKEN}`;
const TELEGRAM_FILE_API = `https://api.telegram.org/file/bot${TELEGRAM_BOT_TOKEN}`;
const VELLUM_EXECUTE_URL = "https://api.vellum.ai/v1/execute-workflow";

async function getTelegramFileUrl(fileId) {
  const resp = await fetch(`${TELEGRAM_API}/getFile?file_id=${fileId}`);
  const data = await resp.json();
  if (data.ok) {
    return `${TELEGRAM_FILE_API}/${data.result.file_path}`;
  }
  return null;
}

async function callVellum(chatId, messageId, userName, text, imageUrl) {
  const payload = {
    workflow_deployment_name: VELLUM_WORKFLOW_NAME,
    inputs: [
      { name: "chat_id", type: "NUMBER", value: chatId },
      { name: "message_id", type: "NUMBER", value: messageId },
      { name: "user_name", type: "STRING", value: userName },
      { name: "message", type: "STRING", value: text || "" },
      { name: "image_url", type: "STRING", value: imageUrl || "" },
    ],
  };

  const resp = await fetch(VELLUM_EXECUTE_URL, {
    method: "POST",
    headers: {
      "X-API-KEY": VELLUM_API_KEY,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!resp.ok) {
    const errorText = await resp.text();
    console.error(`Vellum API error: ${resp.status} — ${errorText}`);
    return null;
  }

  return await resp.json();
}

function extractFileId(message) {
  if (message.document) {
    return message.document.file_id;
  }
  if (message.photo && message.photo.length > 0) {
    return message.photo[message.photo.length - 1].file_id;
  }
  return null;
}

exports.handler = async (event, context) => {
  try {
    const body = JSON.parse(event.body);
    console.log("Received update:", JSON.stringify(body));

    const message = body.message;
    if (!message) {
      return {
        statusCode: 200,
        body: JSON.stringify({ ok: true, detail: "no message" }),
      };
    }

    const chatId = message.chat.id;
    const messageId = message.message_id;
    const userName = message.from.username || message.from.first_name || "unknown";
    const text = message.text || message.caption || "";

    // Resolve image URL if photo/document is attached
    let imageUrl = null;
    const fileId = extractFileId(message);
    if (fileId) {
      imageUrl = await getTelegramFileUrl(fileId);
      console.log("Resolved image URL:", imageUrl);
    }

    // Forward to Vellum (Vellum handles Telegram reply)
    const result = await callVellum(chatId, messageId, userName, text, imageUrl);
    console.log("Vellum response:", JSON.stringify(result));

    return {
      statusCode: 200,
      body: JSON.stringify({ ok: true }),
    };
  } catch (err) {
    console.error("Webhook error:", err);
    return {
      statusCode: 200,
      body: JSON.stringify({ ok: true }),
    };
  }
};

File: netlify/functions/webhook.py
import asyncio
import json
import logging
import os

import httpx
from dotenv import load_dotenv

load_dotenv()

TELEGRAM_BOT_TOKEN = os.getenv("TELEGRAM_BOT_TOKEN")
VELLUM_API_KEY = os.getenv("VELLUM_API_KEY")
VELLUM_WORKFLOW_NAME = os.getenv("VELLUM_WORKFLOW_NAME", "telegram-split-bill-bot")

TELEGRAM_API = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}"
TELEGRAM_FILE_API = f"https://api.telegram.org/file/bot{TELEGRAM_BOT_TOKEN}"
VELLUM_EXECUTE_URL = "https://api.vellum.ai/v1/execute-workflow"

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


async def get_telegram_file_url(file_id: str) -> str | None:
    """Call Telegram getFile API and return the full download URL."""
    async with httpx.AsyncClient(timeout=15) as client:
        resp = await client.get(f"{TELEGRAM_API}/getFile", params={"file_id": file_id})
        data = resp.json()
        if data.get("ok"):
            file_path = data["result"]["file_path"]
            return f"{TELEGRAM_FILE_API}/{file_path}"
    return None


async def call_vellum(chat_id: int, message_id: int, user_name: str, text: str, image_url: str | None):
    """Execute the Vellum workflow with the extracted Telegram data."""
    inputs = [
        {"name": "chat_id", "type": "NUMBER", "value": chat_id},
        {"name": "message_id", "type": "NUMBER", "value": message_id},
        {"name": "user_name", "type": "STRING", "value": user_name},
        {"name": "message", "type": "STRING", "value": text or ""},
        {"name": "image_url", "type": "STRING", "value": image_url or ""},
    ]

    payload = {
        "workflow_deployment_name": VELLUM_WORKFLOW_NAME,
        "inputs": inputs,
    }

    headers = {
        "X-API-KEY": VELLUM_API_KEY,
        "Content-Type": "application/json",
    }

    async with httpx.AsyncClient(timeout=60) as client:
        resp = await client.post(VELLUM_EXECUTE_URL, json=payload, headers=headers)
        resp.raise_for_status()
        return resp.json()


def extract_file_id(message: dict) -> str | None:
    """Extract file_id from document, photo, or compressed image."""
    if "document" in message:
        return message["document"]["file_id"]
    if "photo" in message and message["photo"]:
        return message["photo"][-1]["file_id"]
    return None


async def process_message(chat_id: int, message_id: int, user_name: str, text: str, image_url: str | None):
    """Run Vellum in background (Vellum handles Telegram reply)."""
    try:
        result = await call_vellum(chat_id, message_id, user_name, text, image_url)
        logger.info("Vellum response: %s", result)
    except httpx.HTTPStatusError as e:
        logger.error("Vellum API error: %s — %s", e.response.status_code, e.response.text)
    except Exception as e:
        logger.error("Unexpected error: %s", e)


def handler(event, context):
    """Netlify Function handler for Telegram webhook."""
    try:
        body = json.loads(event["body"]) if isinstance(event["body"], str) else event["body"]
        logger.info("Received update: %s", body)

        message = body.get("message")
        if not message:
            return {"statusCode": 200, "body": json.dumps({"ok": True, "detail": "no message"})}

        chat_id: int = message["chat"]["id"]
        message_id: int = message["message_id"]
        user_name: str = message["from"].get("username") or message["from"].get("first_name", "unknown")
        text: str = message.get("text") or message.get("caption") or ""

        # Resolve image URL if photo/document is attached
        image_url: str | None = None
        file_id = extract_file_id(message)
        if file_id:
            image_url = asyncio.run(get_telegram_file_url(file_id))
            logger.info("Resolved image URL: %s", image_url)

        # Process in background (fire and forget for Netlify)
        asyncio.run(process_message(chat_id, message_id, user_name, text, image_url))

        return {"statusCode": 200, "body": json.dumps({"ok": True})}

    except Exception as e:
        logger.error("Webhook error: %s", e)
        return {"statusCode": 500, "body": json.dumps({"error": str(e)})}
